File: ch14/java/servicebus/ServiceBusExample.java
// Java example: Azure Service Bus - Send and Receive Message
// Requires Maven: com.azure:azure-messaging-servicebus, com.azure:azure-identity

import com.azure.messaging.servicebus.*;
import com.azure.identity.DefaultAzureCredentialBuilder;

public class ServiceBusExample {
    public static void main(String[] args) {
        String queueName = "myqueue";
        String connectionString = "<your_connection_string>";
        String fullyQualifiedNamespace = "<your_servicebus_namespace>.servicebus.windows.net";
        String message = "Hello from Java!";

        // 1. Using connection string
        try (ServiceBusClientBuilder builder = new ServiceBusClientBuilder()) {
            ServiceBusSenderClient sender = builder.connectionString(connectionString).sender().queueName(queueName).buildClient();
            sender.sendMessage(new ServiceBusMessage(message));
            ServiceBusReceiverClient receiver = builder.connectionString(connectionString).receiver().queueName(queueName).buildClient();
            ServiceBusReceivedMessage receivedMessage = receiver.receiveMessages(1).stream().findFirst().orElse(null);
            if (receivedMessage != null)
                System.out.println("Received (conn str): " + receivedMessage.getBody().toString());
            sender.close();
            receiver.close();
        }

        // 2. Using Entra ID (DefaultAzureCredential)
        ServiceBusClientBuilder builderEntra = new ServiceBusClientBuilder()
                .fullyQualifiedNamespace(fullyQualifiedNamespace)
                .credential(new DefaultAzureCredentialBuilder().build());
        ServiceBusSenderClient senderEntra = builderEntra.sender().queueName(queueName).buildClient();
        senderEntra.sendMessage(new ServiceBusMessage(message));
        ServiceBusReceiverClient receiverEntra = builderEntra.receiver().queueName(queueName).buildClient();
        ServiceBusReceivedMessage receivedMessageEntra = receiverEntra.receiveMessages(1).stream().findFirst().orElse(null);
        if (receivedMessageEntra != null)
            System.out.println("Received (Entra ID): " + receivedMessageEntra.getBody().toString());
        senderEntra.close();
        receiverEntra.close();
    }
}

File: ch14/python/servicebus/servicebus_example.py
# Python example: Azure Service Bus - Send and Receive Message
# Requires: pip install azure-servicebus azure-identity

from azure.servicebus import ServiceBusClient, ServiceBusMessage
from azure.identity import DefaultAzureCredential

queue_name = "myqueue"
connection_string = "<your_connection_string>"
fully_qualified_namespace = "<your_servicebus_namespace>.servicebus.windows.net"
message = "Hello from Python!"

# 1. Using connection string
with ServiceBusClient.from_connection_string(connection_string) as client:
    sender = client.get_queue_sender(queue_name)
    with sender:
        sender.send_messages(ServiceBusMessage(message))
    receiver = client.get_queue_receiver(queue_name)
    with receiver:
        for msg in receiver.receive_messages(max_message_count=1, max_wait_time=5):
            print(f"Received (conn str): {str(msg)}")
            receiver.complete_message(msg)

# 2. Using Entra ID (DefaultAzureCredential)
with ServiceBusClient(fully_qualified_namespace, credential=DefaultAzureCredential()) as client:
    sender = client.get_queue_sender(queue_name)
    with sender:
        sender.send_messages(ServiceBusMessage(message))
    receiver = client.get_queue_receiver(queue_name)
    with receiver:
        for msg in receiver.receive_messages(max_message_count=1, max_wait_time=5):
            print(f"Received (Entra ID): {str(msg)}")
            receiver.complete_message(msg)
